### PR TITLE
Add LocaleData parameter for word/sentence segmenter

### PR DESCRIPTION
**Pull request by makotokato**
This is a part of https://github.com/unicode-org/icu4x/issues/3284.

ICU4C has some language break rules for word and sentence segmenter, so this fix adds some rules to ICU4X per locale.

This adds LocaleData argument to all constructors. Since data marker has to remove singleton parameter to support Locale, this happens API breaking change.

### CODE REVIEW

File: provider/data/segmenter/fingerprints.csv [hunk 1]
```diff
@@ -15,5 +15,12 @@ segmenter/lstm/wl_auto@1, und/Burmese_codepoints_exclusive_model4_heavy, 91369B,
 segmenter/lstm/wl_auto@1, und/Khmer_codepoints_exclusive_model4_heavy, 74669B, b25f5219c4b970f2
 segmenter/lstm/wl_auto@1, und/Lao_codepoints_exclusive_model4_heavy, 72164B, 7e0c3ea7801791bd
 segmenter/lstm/wl_auto@1, und/Thai_codepoints_exclusive_model4_heavy, 72331B, c46e2e0c098c1fc1
-segmenter/sentence@2, <singleton>, 14528B, da95c51c93946d95
-segmenter/word@2, <singleton>, 14768B, 7cb863d0724b4119
+segmenter/sentence@2, <lookup>, 41B, 2 identifiers
+segmenter/sentence@2, <total>, 29072B, 2 unique payloads
```
[the remark this answers is not in the document]
Comment (sffc) [reply provider/data/segmenter/fingerprints.csv:19]: +1, duplicating the data doesn't scale; having an extension/supplement that contains the diff from the standard payload scales better if many locales have diffs.

[the remark this answers is not in the document]
Comment (makotokato) [reply provider/data/segmenter/fingerprints.csv:19]: @robertbastian, should we keep singleton then should we have both property data into UND? 
The different is property table only (codepoint to property). And rule table for state machine is same.

File: components/segmenter/src/lib.rs [hunk 2]
```diff
@@ -67,9 +67,10 @@
 //! Find all word boundaries:
 //!
 //!```rust
+//! use icu::locale::locale;
 //! use icu::segmenter::WordSegmenter;
 //!
-//! let segmenter = WordSegmenter::new_auto();
+//! let segmenter = WordSegmenter::try_new_auto(&locale!("en").into()).unwrap();
```
Comment (sffc) [inline components/segmenter/src/lib.rs:73]: Note: The request was for this to be

`new_auto()`

along with

`try_new_auto_with_options(WordSegmenterOptions::new_with_content_locale(&locale!("en")))`

File: components/segmenter/tests/locale.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_locale_core::locale;
+use icu_segmenter::{SentenceSegmenter, WordSegmenter};
+
+// Additional segmenter tests with locale.
+
+#[test]
+fn word_break_with_locale() {
```
Comment (sffc) [inline components/segmenter/tests/locale.rs:11]: Praise: Glad these cases are working now

**Review by sffc — CHANGES_REQUESTED** [review]
Good that there are cases that work now that failed before. But, the consensus was to implement this while still keeping the previous, infallible, singleton constructors.

**Comment by makotokato** [comment]
I will update this fix with singleton by adding the difference of property_table to RuleBreakDataV2.

**Comment by robertbastian** [comment]
Please no force pushes after a started review, [this](https://github.com/unicode-org/icu4x/compare/dd6855da75b2ec0bc7a95cf73f77a3f58e42c570..5adb60e2c2c637a7b22a7e941ee306d58134bbcd) is what the diff looks like now.

**Comment by makotokato** [comment]
It might be better to add new data marker for locale diff instead, so I will update this

**Comment by makotokato** [comment]
> Please no force pushes after a started review, [this](https://github.com/unicode-org/icu4x/compare/dd6855da75b2ec0bc7a95cf73f77a3f58e42c570..5adb60e2c2c637a7b22a7e941ee306d58134bbcd) is what the diff looks like now.

I change my idea, not use simply locale, so I would like to reset all.  Should I close this PR, then send new PR instead?

**Comment by sffc** [comment]
Yes, if the change is sufficiently different, please open a new PR.

**Comment by makotokato** [comment]
OK. I close this PR, then file new PR

**Comment by makotokato** [comment]
New PR is https://github.com/unicode-org/icu4x/pull/5318

**Pull request closed without merging**